dimod/constrained/cyexpression.pyx

```diff
@@ -362,16 +362,73 @@ cdef class cyConstraintView(_cyExpression):
         self.constraint().mark_discrete(marker)
 
     def penalty(self):
-        penalty = self.constraint().penalty()
+        """The penalty type for a soft constraint.
 
-        if penalty == cppPenalty.LINEAR:
+        Returns:
+            If the constraint is soft, returns one of ``'linear'`` or ``'quadratic'``.
+            If the constraint is hard, returns ``None``.
+
+        """
+        constraint = self.constraint()
+        penalty = constraint.penalty()
+
+        if not constraint.is_soft():
+            return None
+        elif penalty == cppPenalty.LINEAR:
             return "linear"
         elif penalty == cppPenalty.QUADRATIC:
             return "quadratic"
         elif penalty == cppPenalty.CONSTANT:
-            return "constrant"
+            # user should never see this, but might as well be future proof
+            return "constant"
         else:
             raise RuntimeError("unexpected penalty")
 
+    def set_weight(self, weight, penalty='linear'):
```

Comment:
Perhaps use enums for penalty type?

Reply:
I would love to, unfortunately to maintain compatibility with 0.11.6 we're stuck with strings. Could go through a deprecation period but...